modules/apps/dynamic-data-mapping/dynamic-data-mapping-form-field-type/src/main/resources/META-INF/resources/ObjectRelationship/ObjectRelationship.js

```diff
@@ -12,33 +12,101 @@
  * details.
  */
 
-import Autocomplete from 'commerce-frontend-js/components/autocomplete/Autocomplete';
+import {useIsMounted} from '@liferay/frontend-js-react-web';
 import {FieldBase} from 'dynamic-data-mapping-form-field-type/FieldBase/ReactFieldBase.es';
-import React from 'react';
+import {Autocomplete as BaseAutocomplete} from 'frontend-js-components-web';
+import {debounce, fetch} from 'frontend-js-web';
+import React, {useState} from 'react';
 
-const ObjectRelationship = ({
+function getData(apiUrl, query, page, pageSize) {
+	const url = new URL(apiUrl, themeDisplay.getPortalURL());
+
+	if (query) {
+		url.searchParams.set('search', query);
+	}
+
+	if (page) {
+		url.searchParams.set('page', page);
+	}
+
+	if (pageSize) {
+		url.searchParams.set('pageSize', pageSize);
+	}
+
+	return fetch(url, {
+		headers: new Headers({
+			Accept: 'application/json',
+			'Accept-Language': themeDisplay.getBCP47LanguageId(),
+			'Content-Type': 'application/json',
+		}),
+	}).then((data) => data.json());
+}
+
+export function ObjectRelationship({
 	apiURL,
 	initialLabel,
 	initialValue,
 	inputName,
-	itemsKey,
-	itemsLabel,
+	labelKey,
 	name,
 	value,
+	valueKey,
 	...otherProps
-}) => (
-	<FieldBase name={name} {...otherProps}>
-		<Autocomplete
-			apiUrl={apiURL}
-			initialLabel={initialLabel}
-			initialValue={initialValue}
-			inputName={inputName}
-			itemsKey={itemsKey}
-			itemsLabel={itemsLabel}
-			name={name}
-			value={value}
-		/>
-	</FieldBase>
-);
+}) {
+	const [query, setQuery] = useState(initialLabel || '');
+	const [selectedItem, setSelectedItem] = useState(initialValue || value);
+	const [items, setItems] = useState(null);
+	const [loading, setLoading] = useState(false);
+	const isMounted = useIsMounted();
+
+	const fetchData = debounce(() => {
+		if (isMounted()) {
+			setLoading(true);
+
+			getData(apiURL, query, 1, 10)
+				.then((jsonResponse) => {
+					setItems(jsonResponse.items);
+
+					setLoading(false);
+
+					if (!query) {
+						return;
+					}
+
+					const found = jsonResponse.items.find(
+						(item) => item[labelKey] === query
+					);
+
+					if (found) {
+						setSelectedItem(found);
+					}
+				})
+				.catch(() => {
+					setLoading(false);
+				});
+		}
+	}, 500);
+
+	return (
+		<FieldBase name={name} {...otherProps}>
+			<BaseAutocomplete
+				inputName={inputName}
+				inputValue={query}
+				items={items}
+				labelKey={labelKey}
+				loading={loading}
+				name={name}
+				onInputChange={(val) => {
+					setQuery(val);
+					setSelectedItem(null);
+					fetchData();
```

Comment:
This is pretty weird, I think `fetchData` is working exactly because of `debounce`, because we set the `query` here and it's asynchronous, so we call `fetchData` right away.

Reply:
Oh good point, I thought something seemed off when I was writing this but I just ignored the feeling 😂

Reply:
haha 😂